server/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -274,8 +275,26 @@ public synchronized void applyClusterState(final ClusterChangedEvent event) {
         lastClusterStateShardsClosedListener = new SubscribableListener<>();
         currentClusterStateShardsClosedListeners = new RefCountingListener(lastClusterStateShardsClosedListener);
         try {
-            previousShardsClosedListener.addListener(currentClusterStateShardsClosedListeners.acquire());
```

Comment:
Hmm are you sure we should move all this listener stuff below `doApplyClusterState()`?

Reply:
But I've put it back at the original place.